homeassistant/components/mobile_app/__init__.py

```diff
@@ -0,0 +1,396 @@
+        return None
+
+    key = key.encode("utf-8")
+    key = key[:keylen]
+    key = key.ljust(keylen, b'\0')
+
+    try:
+        message = decrypt(ciphertext, key)
+        message = json.loads(message.decode("utf-8"))
+        _LOGGER.debug("Successfully decrypted mobile_app payload")
+        return message
+    except ValueError:
+        _LOGGER.warning("Ignoring encrypted payload because unable to decrypt")
+        return None
+
+
+def context(request):
+    """Generate a context from a request."""
+    user = request.get('hass_user')
+    if user is None:
+        return Context()
+
+    return Context(user_id=user.id)
+
+
+async def handle_webhook(hass: HomeAssistantType, webhook_id: str, request):
+    """Handle webhook callback."""
+    device = device_for_webhook_id(hass, webhook_id)
+
+    try:
+        req_data = await request.json()
+    except ValueError:
+        _LOGGER.warning('Received invalid JSON from mobile_app')
+        return json_response([], status=HTTP_BAD_REQUEST)
+
+    try:
+        req_data = WEBHOOK_PAYLOAD_SCHEMA(req_data)
+    except vol.Invalid as ex:
+        return json_response(vol.humanize.humanize_error(req_data, ex),
+                             status=HTTP_BAD_REQUEST)
+
+    webhook_type = req_data[ATTR_WEBHOOK_TYPE]
+
+    webhook_payload = req_data.get(ATTR_WEBHOOK_DATA, {})
+
+    if req_data[ATTR_WEBHOOK_ENCRYPTED]:
+        enc_data = req_data[ATTR_WEBHOOK_ENCRYPTED_DATA]
+        webhook_payload = _decrypt_payload(device[CONF_SECRET], enc_data)
+
+    try:
+        data = WEBHOOK_SCHEMAS[webhook_type](webhook_payload)
+    except vol.Invalid as ex:
+        return json_response(vol.humanize.humanize_error(webhook_payload, ex),
+                             status=HTTP_BAD_REQUEST)
+
+    if webhook_type == WEBHOOK_TYPE_CALL_SERVICE:
+
+        try:
+            await hass.services.async_call(data[ATTR_DOMAIN],
+                                           data[ATTR_SERVICE],
+                                           data[ATTR_SERVICE_DATA],
+                                           True,
+                                           context(request))
+        except (vol.Invalid, ServiceNotFound):
+            raise HTTPBadRequest()
+
+        return Response(status=200)
+
+    elif webhook_type == WEBHOOK_TYPE_FIRE_EVENT:
+        event_type = data[ATTR_EVENT_TYPE]
+        hass.bus.async_fire(event_type, data[ATTR_EVENT_DATA],
+                            ha.EventOrigin.remote, context(request))
+        return Response(status=200)
+
+    elif webhook_type == WEBHOOK_TYPE_RENDER_TEMPLATE:
+        try:
+            tpl = template.Template(data[ATTR_TEMPLATE], hass)
+            rendered = tpl.async_render(data.get(ATTR_TEMPLATE_VARIABLES))
+            return json_response({"rendered": rendered})
+        except (ValueError, TemplateError) as ex:
+            return json_response("Error rendering template: {}".format(ex),
```

Comment:
Same: let's log but return 200.

Reply:
This one I have a bit of an issue with changing, since the template could be malformed _or_ the variables could have an issue and the app would never know. We/I don't want to implement a whole Jinja parser into my app to ensure that it's valid before sending.

Reply:
ah right. Let's keep it in . But let's not just do a json response with a string, wrap it in a dict.